Add edit profile flow

ahmarzoukk opened the pull request:
This PR handles issue https://github.com/yebrahim/codersquare/issues/30. It mainly does the following: 

- Adds a backend endpoint for updating profiles:  `PATCH api/v1/users`.  

It requires the following in the request body:
```json
{
 "userName": "string",
 "firstName": "string",
 "lastName": "string",
}
```
Users can update any of the above-mentioned attributes and only theirs as the updates reference the already logged-in user. 

- Updates the UI slightly to allow users to edit each of the previously mentioned profile attributes. 

The new UI changes are screenshotted below based on the related use cases:

Users viewing other users' profiles:
![Users viewing other users' profiles](https://user-images.githubusercontent.com/104036623/198388060-a72ced3c-290a-4c5f-b0af-ff7a0ff29fe7.png)
Users viewing their own profiles:
![Users viewing their own profile  ](https://user-images.githubusercontent.com/104036623/198388249-d1bf12a8-534e-4cc5-b7a2-5ff611aed6e8.png)
Users editing their own profiles:
![Users editing their own profile  ](https://user-images.githubusercontent.com/104036623/198388177-c18bd1c5-d5a7-45f7-a165-d238e31abcdf.png)

- Updates the build script at the root package.json to install the dev dependencies that it uses (e.g. `npm-run-all`)

yebrahim commented on web/src/components/navbar.tsx:
```diff
@@ -19,11 +20,16 @@ export const Navbar = () => {
     signOut();
```
need to also set current user to undefined here.

yebrahim commented on web/src/App.tsx:
```diff
@@ -11,22 +13,32 @@ import { ViewPost } from './pages/view-post';
 import { isDev } from './util';
 
+type UserContext = {
+  currentUser: GetCurrentUserResponse;
+  setCurrentUserGlobally: Dispatch<SetStateAction<GetCurrentUserResponse>>;
```
these should probably be optional?
```
{
  currentUser?: ...
  setCurrentUserGlobally?: ...
}
```
then you could use the type `UserContext` when creating the state below: `useState<UserContext>({});`.

ahmarzoukk replied:
This didn't work for me as I expect.

yebrahim commented on web/src/pages/user-profile.tsx:
```diff
@@ -1,16 +1,66 @@
+            lastName,
+          }
+        );
+        window.location.reload();
```
instead of a full reload, we could just call the `setCurrentUserGlobally` callback from the context?

yebrahim replied:
also `setUserUpdateError('')` in case of success here.

yebrahim commented on server/src/handlers/userHandler.ts:
```diff
@@ -114,6 +116,26 @@ export class UserHandler {
     });
   };
 
+  public updateCurrentUser: ExpressHandler<UpdateCurrentUserRequest, UpdateCurrentUserResponse> =
```
this endpoint allows you to edit other users' username, no?
we need another check that the `existingUser`'s id below is equal to `res.locals.userId`.

ahmarzoukk replied:
No, it doesn't. All profile updates target the current authenticated use as I do the updates query against `res.locals.userId`.

yebrahim reviewed (commented):
thanks for the PR, and sorry for the late review! I appreciate the screenshot as well! :)

ahmarzoukk commented:
@yebrahim Sorry I got busy lately. Hopefully, I'll be able to address your comments this weeekend.

yebrahim commented:
@ahmarzoukk no worries, thanks for the contribution. :)

yebrahim commented:
@ahmarzoukk are you getting back. on this or should I just commandeer it? :)

ahmarzoukk commented:
@yebrahim This PR had entirely slipped my mind. So sorry for any inconvenience! 

I tried to improve my changes a little bit while I was addressing your comments! Please review!

yebrahim commented on web/src/components/userContext.tsx:
```diff
@@ -0,0 +1,39 @@
+
+type UserContext = {
+  currentUser?: GetCurrentUserResponse;
+  refreshCurrentUser?: any;
```
nit: type this as a function, maybe something like `refreshCurrentUser?: () => void`.

yebrahim reviewed (approved):
nice, thanks for getting back on this. lgtm overall, with a few nits but I won't block on them. let me know if you wanna merge as is. :)

ahmarzoukk commented:
@yebrahim, I addressed your comments. Please feel free to merge once you're happy with the changes.
